File: smart_screens/smart_screens/page/batch_conversion_entry/batch_conversion_entry.js
frappe.pages['batch-conversion-entry'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Batch Conversion Entry',
		single_column: true
	});

	// Directly set the HTML content
	page.main.html(`
		<div class="batch-conversion-entry-page" style="padding: 20px;">
			<div class="page-header">
				<h3>Batch Conversion Data Entry</h3>
				<p class="text-muted">Manually enter moulding batch conversion data month by month</p>
			</div>

			<div class="alert alert-info">
				<strong><i class="fa fa-info-circle"></i> Purpose:</strong> 
				Capture mould specification and conversion factors for finished goods batches to enable kg→nos conversion in aggregated reports.
			</div>

			<div class="filter-section" style="margin-bottom: 20px; padding: 15px; background: #f9f9f9; border-radius: 5px;">
				<div class="row">
					<div class="col-md-3">
						<label>Select Month</label>
						<input type="month" id="select_month" class="form-control" />
					</div>
					<div class="col-md-3">
						<label>Company</label>
						<input type="text" id="company" class="form-control" readonly value="SPP INDIA" />
					</div>
					<div class="col-md-3">
						<label>&nbsp;</label>
						<button class="btn btn-primary btn-block" id="btn_load_batches">
							<i class="fa fa-search"></i> Load Batches
						</button>
					</div>
					<div class="col-md-3">
						<label>&nbsp;</label>
						<button class="btn btn-success btn-block" id="btn_process_month" disabled>
							<i class="fa fa-cogs"></i> Process Month (Background)
						</button>
					</div>
				</div>
			</div>

			<div class="stats-section" style="margin-bottom: 20px;">
				<div class="row">
					<div class="col-md-3">
						<div class="stat-box" style="padding: 15px; background: #e3f2fd; border-radius: 5px; text-align: center;">
							<h4 id="stat_total">0</h4>
							<small>Total Batches</small>
						</div>
					</div>
					<div class="col-md-3">
						<div class="stat-box" style="padding: 15px; background: #c8e6c9; border-radius: 5px; text-align: center;">
							<h4 id="stat_processed">0</h4>
							<small>Already Processed</small>
						</div>
					</div>
					<div class="col-md-3">
						<div class="stat-box" style="padding: 15px; background: #fff3e0; border-radius: 5px; text-align: center;">
							<h4 id="stat_pending">0</h4>
							<small>Pending</small>
						</div>
					</div>
					<div class="col-md-3">
						<div class="stat-box" style="padding: 15px; background: #ffebee; border-radius: 5px; text-align: center;">
							<h4 id="stat_errors">0</h4>
							<small>Errors</small>
						</div>
					</div>
				</div>
			</div>

			<div class="batch-list-section">
				<h4>Batch Details</h4>
				<div id="batch_table_container">
					<p class="text-muted">Select a month and click "Load Batches" to view production entries.</p>
				</div>
			</div>

			<div class="background-job-section" style="margin-top: 30px; padding: 15px; background: #f5f5f5; border-radius: 5px; display: none;" id="job_status_section">
				<h4>Background Job Status</h4>
				<div class="progress" style="height: 25px;">
					<div class="progress-bar progress-bar-striped active" id="job_progress" role="progressbar" style="width: 0%;">
						<span id="progress_text">0%</span>
					</div>
				</div>
				<p class="text-muted" style="margin-top: 10px;" id="job_message">Waiting to start...</p>
			</div>
		</div>
	`);

	setup_page(wrapper, page);
};

function setup_page(wrapper, page) {
	let current_month_data = [];
	
	// Set default month to current month
	const today = new Date();
	const currentMonth = today.toISOString().slice(0, 7);
	$(wrapper).find('#select_month').val(currentMonth);

	// Load Batches button
	$(wrapper).find('#btn_load_batches').on('click', function() {
		const selected_month = $(wrapper).find('#select_month').val();
		if (!selected_month) {
			frappe.msgprint('Please select a month');
			return;
		}
		load_month_batches(wrapper, selected_month);
	});

	// Process Month button
	$(wrapper).find('#btn_process_month').on('click', function() {
		const selected_month = $(wrapper).find('#select_month').val();
		if (!selected_month) {
			frappe.msgprint('Please select a month');
			return;
		}
		process_month_background(wrapper, selected_month);
	});

	function load_month_batches(wrapper, month) {
		$(wrapper).find('#batch_table_container').html(`
			<div class="text-center" style="padding: 40px;">
				<i class="fa fa-spinner fa-spin fa-2x text-muted"></i>
				<p class="text-muted" style="margin-top: 10px;">Loading batches for ${month}...</p>
			</div>
		`);

		frappe.call({
			method: 'smart_screens.smart_screens.page.batch_conversion_entry.batch_conversion_entry.get_month_batches',
			args: { month: month },
			callback: function(r) {
				if (r.message && r.message.success) {
					current_month_data = r.message.data;
					render_batch_table(wrapper, r.message.data, r.message.stats);
					$(wrapper).find('#btn_process_month').prop('disabled', false);
				} else {
					show_error(wrapper, r.message?.error || 'Failed to load batches');
				}
			}
		});
	}

	function render_batch_table(wrapper, data, stats) {
		// Update stats
		$(wrapper).find('#stat_total').text(stats.total);
		$(wrapper).find('#stat_processed').text(stats.processed);
		$(wrapper).find('#stat_pending').text(stats.pending);
		$(wrapper).find('#stat_errors').text(stats.errors);

		if (!data || data.length === 0) {
			$(wrapper).find('#batch_table_container').html(`
				<div class="alert alert-warning">No production entries found for this month.</div>
			`);
			return;
		}

		let html = `
			<div class="table-responsive">
				<table class="table table-bordered table-striped table-hover">
					<thead>
						<tr>
							<th>MPE</th>
							<th>Date</th>
							<th>Finished Item</th>
							<th>Finished Batch</th>
							<th>Weight (kg)</th>
							<th>Lifts</th>
							<th>Cavities</th>
							<th>Conversion Factor</th>
							<th>Status</th>
							<th>Action</th>
						</tr>
					</thead>
					<tbody>
		`;

		data.forEach(row => {
			const conversion_factor = row.no_of_running_cavities && row.number_of_lifts && row.weight > 0
				? ((row.no_of_running_cavities * row.number_of_lifts) / row.weight).toFixed(4)
				: 'N/A';
			
			const status_badge = row.is_processed 
				? '<span class="label label-success">Processed</span>'
				: '<span class="label label-warning">Pending</span>';

			html += `
				<tr>
					<td><a href="/app/moulding-production-entry/${row.mpe_name}" target="_blank">${row.mpe_name}</a></td>
					<td>${frappe.datetime.str_to_user(row.posting_date)}</td>
					<td>${row.finished_item || 'N/A'}</td>
					<td><strong>${row.finished_batch || 'N/A'}</strong></td>
					<td class="text-right">${row.weight.toFixed(3)}</td>
					<td class="text-right">${row.number_of_lifts}</td>
					<td class="text-right">${row.no_of_running_cavities}</td>
					<td class="text-right"><strong>${conversion_factor}</strong></td>
					<td>${status_badge}</td>
					<td>
						${!row.is_processed ? 
							`<button class="btn btn-xs btn-primary btn-process-single" data-mpe="${row.mpe_name}">
								<i class="fa fa-check"></i> Process
							</button>` : 
							`<button class="btn btn-xs btn-default" disabled>Done</button>`
						}
					</td>
				</tr>
			`;
		});

		html += `</tbody></table></div>`;

		$(wrapper).find('#batch_table_container').html(html);

		// Bind single process buttons
		$(wrapper).find('.btn-process-single').on('click', function() {
			const mpe_name = $(this).data('mpe');
			process_single_batch(wrapper, mpe_name);
		});
	}

	function process_single_batch(wrapper, mpe_name) {
		frappe.call({
			method: 'smart_screens.smart_screens.page.batch_conversion_entry.batch_conversion_entry.process_single_batch',
			args: { mpe_name: mpe_name },
			callback: function(r) {
				if (r.message && r.message.success) {
					frappe.show_alert({
						message: `Processed ${mpe_name} successfully`,
						indicator: 'green'
					});
					// Reload the table
					const selected_month = $(wrapper).find('#select_month').val();
					load_month_batches(wrapper, selected_month);
				} else {
					frappe.msgprint({
						title: 'Error',
						message: r.message?.error || 'Failed to process batch',
						indicator: 'red'
					});
				}
			}
		});
	}

	function process_month_background(wrapper, month) {
		frappe.confirm(
			`This will process all pending batches for ${month} in the background. Continue?`,
			function() {
				$(wrapper).find('#job_status_section').show();
				$(wrapper).find('#btn_process_month').prop('disabled', true);

				frappe.call({
					method: 'smart_screens.smart_screens.page.batch_conversion_entry.batch_conversion_entry.process_month_background',
					args: { month: month },
					callback: function(r) {
						if (r.message && r.message.success) {
							frappe.show_alert({
								message: 'Background job started successfully',
								indicator: 'green'
							});
							monitor_background_job(wrapper, r.message.job_id, month);
						} else {
							frappe.msgprint({
								title: 'Error',
								message: r.message?.error || 'Failed to start background job',
								indicator: 'red'
							});
							$(wrapper).find('#btn_process_month').prop('disabled', false);
						}
					}
				});
			}
		);
	}

	function monitor_background_job(wrapper, job_id, month) {
		const interval = setInterval(function() {
			frappe.call({
				method: 'smart_screens.smart_screens.page.batch_conversion_entry.batch_conversion_entry.get_job_status',
				args: { job_id: job_id },
				callback: function(r) {
					if (r.message) {
						const progress = r.message.progress || 0;
						$(wrapper).find('#job_progress').css('width', progress + '%');
						$(wrapper).find('#progress_text').text(progress + '%');
						$(wrapper).find('#job_message').text(r.message.message || 'Processing...');

						if (r.message.status === 'completed') {
							clearInterval(interval);
							$(wrapper).find('#job_progress').removeClass('active');
							frappe.show_alert({
								message: 'Background job completed successfully!',
								indicator: 'green'
							});
							// Reload batches
							setTimeout(function() {
								load_month_batches(wrapper, month);
								$(wrapper).find('#job_status_section').hide();
							}, 2000);
						} else if (r.message.status === 'failed') {
							clearInterval(interval);
							$(wrapper).find('#job_progress').removeClass('active').addClass('progress-bar-danger');
							frappe.msgprint({
								title: 'Job Failed',
								message: r.message.error || 'Background job failed',
								indicator: 'red'
							});
						}
					}
				}
			});
		}, 2000); // Check every 2 seconds
	}

	function show_error(wrapper, message) {
		$(wrapper).find('#batch_table_container').html(`
			<div class="alert alert-danger">${message}</div>
		`);
	}
}

File: smart_screens/smart_screens/page/batch_conversion_entry/batch_conversion_entry.py
# Copyright (c) 2025, Aggregated Report and contributors
# For license information, please see license.txt

import frappe
from frappe import _
import json
from datetime import datetime
import uuid

@frappe.whitelist()
def get_month_batches(month):
	"""
	Get all Moulding Production Entries for a specific month with their conversion status
	
	Args:
		month: YYYY-MM format (e.g., '2025-11')
	
	Returns:
		{
			success: bool,
			data: list of batch records,
			stats: {total, processed, pending, errors}
		}
	"""
	try:
		# Parse month
		year, month_num = month.split('-')
		
		# Query to get production entries with stock entry details
		query = """
			SELECT 
				mpe.name as mpe_name,
				mpe.moulding_date,
				mpe.batch_no as raw_material_batch,
				mpe.weight,
				mpe.number_of_lifts,
				mpe.no_of_running_cavities,
				mpe.mould_reference,
				mpe.stock_entry_reference,
				se.posting_date,
				se.company,
				sed.item_code as finished_item,
				sed.qty as finished_qty_kg,
				sed.batch_no as finished_batch,
				sed.t_warehouse,
				CASE 
					WHEN mbc.name IS NOT NULL THEN 1 
					ELSE 0 
				END as is_processed
			FROM `tabMoulding Production Entry` mpe
			LEFT JOIN `tabStock Entry` se ON se.name = mpe.stock_entry_reference
			LEFT JOIN `tabStock Entry Detail` sed ON sed.parent = se.name AND sed.is_finished_item = 1
			LEFT JOIN `tabMoulding Batch Conversion` mbc ON mbc.production_entry = mpe.name
			WHERE mpe.docstatus = 1
			AND YEAR(se.posting_date) = %s
			AND MONTH(se.posting_date) = %s
			ORDER BY se.posting_date ASC, mpe.name ASC
		"""
		
		data = frappe.db.sql(query, (year, month_num), as_dict=True)
		
		# Calculate stats
		total = len(data)
		processed = sum(1 for row in data if row.get('is_processed'))
		pending = total - processed
		errors = 0  # Will be calculated based on missing data
		
		# Check for errors (missing critical data)
		for row in data:
			if not row.get('finished_batch') or not row.get('finished_item'):
				errors += 1
		
		stats = {
			'total': total,
			'processed': processed,
			'pending': pending,
			'errors': errors
		}
		
		return {
			'success': True,
			'data': data,
			'stats': stats
		}
		
	except Exception as e:
		frappe.log_error(f"Error in get_month_batches: {str(e)}", "Batch Conversion Entry Error")
		return {
			'success': False,
			'error': str(e)
		}


@frappe.whitelist()
def process_single_batch(mpe_name):
	"""
	Process a single Moulding Production Entry and create Batch Conversion record
	
	Args:
		mpe_name: Name of Moulding Production Entry
	
	Returns:
		{success: bool, error: str}
	"""
	try:
		# Check if already processed
		if frappe.db.exists("Moulding Batch Conversion", {"production_entry": mpe_name}):
			return {
				'success': False,
				'error': f'Batch conversion already exists for {mpe_name}'
			}
		
		# Get Moulding Production Entry
		mpe = frappe.get_doc("Moulding Production Entry", mpe_name)
		
		if not mpe.stock_entry_reference:
			return {
				'success': False,
				'error': f'No stock entry reference found for {mpe_name}'
			}
		
		# Get Stock Entry
		se = frappe.get_doc("Stock Entry", mpe.stock_entry_reference)
		
		# Get finished item details
		finished_item = None
		for item in se.items:
			if item.is_finished_item:
				finished_item = item
				break
		
		if not finished_item:
			return {
				'success': False,
				'error': f'No finished item found in Stock Entry {se.name}'
			}
		
		# Find Mould Specification by mould_ref field (not by name)
		mould_spec_name = None
		avg_blank_weight = None
		if mpe.mould_reference:
			mould_spec_data = frappe.db.get_value(
				"Mould Specification",
				{"mould_ref": mpe.mould_reference},
				["name", "avg_blank_wtproduct_gms"],
				as_dict=True
			)
			if mould_spec_data:
				mould_spec_name = mould_spec_data.get("name")
				avg_blank_weight = mould_spec_data.get("avg_blank_wtproduct_gms")
		
		# Create Moulding Batch Conversion record
		conversion_doc = frappe.get_doc({
			"doctype": "Moulding Batch Conversion",
			"production_entry": mpe.name,
			"batch_no": finished_item.batch_no,
			"item_code": finished_item.item_code,
			"warehouse": finished_item.t_warehouse,
			"posting_datetime": se.posting_date,
			"cavities_per_cycle": mpe.no_of_running_cavities,
			"cycle_time_sec": mpe.curing_time if hasattr(mpe, 'curing_time') else None,
			"shots": mpe.number_of_lifts,
			"gross_qty_kg": mpe.weight,
			"net_qty_kg": finished_item.qty,
			"rejects_qty_kg": 0,  # Calculate if available
			"company": se.company,
			"status": "Active"
		})
		
		 # Set mould_spec if found
		if mould_spec_name:
			conversion_doc.mould_spec = mould_spec_name
		
		conversion_doc.insert(ignore_permissions=True)
		frappe.db.commit()
		
		return {
			'success': True,
			'conversion_doc': conversion_doc.name
		}
		
	except Exception as e:
		frappe.log_error(
			f"Error processing {mpe_name}: {str(e)}\n{frappe.get_traceback()}",
			"Batch Conversion Processing Error"
		)
		return {
			'success': False,
			'error': str(e)
		}


@frappe.whitelist()
def process_month_background(month):
	"""
	Start a background job to process all pending batches for a month
	
	Args:
		month: YYYY-MM format
	
	Returns:
		{success: bool, job_id: str}
	"""
	try:
		# Generate unique job ID
		job_id = str(uuid.uuid4())
		
		# Enqueue background job
		frappe.enqueue(
			method='aggregated_report.aggregated_report.page.batch_conversion_entry.batch_conversion_entry.process_month_job',
			queue='long',
			timeout=3600,
			job_id=job_id,
			month=month,
			job_id_key=job_id
		)
		
		# Store job status in cache
		frappe.cache().set_value(
			f'batch_conversion_job_{job_id}',
			{
				'status': 'started',
				'progress': 0,
				'message': 'Job started...',
				'month': month
			},
			expires_in_sec=3600
		)
		
		return {
			'success': True,
			'job_id': job_id
		}
		
	except Exception as e:
		frappe.log_error(f"Error starting background job: {str(e)}", "Batch Conversion Job Error")
		return {
			'success': False,
			'error': str(e)
		}


@frappe.whitelist()
def get_job_status(job_id):
	"""
	Get the status of a background job
	
	Args:
		job_id: UUID of the job
	
	Returns:
		{status: str, progress: int, message: str}
	"""
	try:
		status = frappe.cache().get_value(f'batch_conversion_job_{job_id}')
		
		if not status:
			return {
				'status': 'not_found',
				'progress': 0,
				'message': 'Job not found'
			}
		
		return status
		
	except Exception as e:
		return {
			'status': 'error',
			'progress': 0,
			'message': str(e)
		}


def process_month_job(month, job_id_key):
	"""
	Background job to process all batches for a month
	
	Args:
		month: YYYY-MM format
		job_id_key: Job ID for status tracking
	"""
	try:
		# Update status
		update_job_status(job_id_key, 'running', 0, f'Loading batches for {month}...')
		
		# Get all pending batches
		result = get_month_batches(month)
		
		if not result.get('success'):
			update_job_status(job_id_key, 'failed', 0, f'Failed to load batches: {result.get("error")}')
			return
		
		data = result.get('data', [])
		pending_batches = [row for row in data if not row.get('is_processed')]
		
		total = len(pending_batches)
		if total == 0:
			update_job_status(job_id_key, 'completed', 100, 'No pending batches to process')
			return
		
		processed = 0
		failed = 0
		
		# Process each batch
		for idx, batch in enumerate(pending_batches, 1):
			try:
				update_job_status(
					job_id_key, 
					'running', 
					int((idx / total) * 100), 
					f'Processing {idx}/{total}: {batch["mpe_name"]}...'
				)
				
				result = process_single_batch(batch['mpe_name'])
				
				if result.get('success'):
					processed += 1
				else:
					failed += 1
					frappe.log_error(
						f"Failed to process {batch['mpe_name']}: {result.get('error')}",
						"Batch Conversion Job Error"
					)
				
				# Commit every 10 records
				if idx % 10 == 0:
					frappe.db.commit()
				
			except Exception as e:
				failed += 1
				frappe.log_error(
					f"Error processing {batch['mpe_name']}: {str(e)}\n{frappe.get_traceback()}",
					"Batch Conversion Job Error"
				)
		
		# Final commit
		frappe.db.commit()
		
		# Update final status
		update_job_status(
			job_id_key, 
			'completed', 
			100, 
			f'Completed! Processed: {processed}, Failed: {failed}, Total: {total}'
		)
		
	except Exception as e:
		frappe.log_error(
			f"Error in process_month_job: {str(e)}\n{frappe.get_traceback()}",
			"Batch Conversion Job Error"
		)
		update_job_status(job_id_key, 'failed', 0, f'Job failed: {str(e)}')


def update_job_status(job_id, status, progress, message):
	"""Update job status in cache"""
	frappe.cache().set_value(
		f'batch_conversion_job_{job_id}',
		{
			'status': status,
			'progress': progress,
			'message': message,
			'updated_at': datetime.now().isoformat()
		},
		expires_in_sec=3600
	)
